File: java/starter/template/src/main/java/network/t0/provider/handler/PaymentHandler.java
package network.t0.provider.handler;

import io.grpc.stub.StreamObserver;
import network.t0.sdk.proto.tzero.v1.common.PaymentReceipt;
import network.t0.sdk.proto.tzero.v1.payment.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Implementation of the ProviderService.
 *
 * <p>Please refer to docs, proto definition comments or source code comments
 * to understand purpose of functions.
 */
public class PaymentHandler extends ProviderServiceGrpc.ProviderServiceImplBase {

    private static final Logger log = LoggerFactory.getLogger(PaymentHandler.class);

    private final NetworkServiceGrpc.NetworkServiceBlockingStub networkClient;

    public PaymentHandler(NetworkServiceGrpc.NetworkServiceBlockingStub networkClient) {
        this.networkClient = networkClient;
    }

    // TODO: Step 2.1 implement how you handle updates of payment initiated by you
    @Override
    public void updatePayment(UpdatePaymentRequest request, StreamObserver<UpdatePaymentResponse> responseObserver) {
        log.info("Received updatePayment for payment_id={}, client_id={}",
                request.getPaymentId(), request.getPaymentClientId());

        // Handle the payment update based on the result type
        switch (request.getResultCase()) {
            case ACCEPTED -> {
                log.info("Payment {} was accepted", request.getPaymentId());
                // TODO: Update your system to reflect the accepted status
            }
            case FAILED -> {
                log.warn("Payment {} failed", request.getPaymentId());
                // TODO: Update your system to reflect the failed status
            }
            case CONFIRMED -> {
                log.info("Payment {} was confirmed/completed", request.getPaymentId());
                // TODO: Update your system to reflect the completed status
            }
            case MANUAL_AML_CHECK -> {
                log.info("Payment {} requires manual AML check", request.getPaymentId());
                // TODO: Handle manual AML check requirement
            }
            default -> log.warn("Unknown result type for payment {}", request.getPaymentId());
        }

        responseObserver.onNext(UpdatePaymentResponse.newBuilder().build());
        responseObserver.onCompleted();
    }

    // TODO: Step 2.4 implement how you do payouts (payments initiated by your counterparts)
    @Override
    public void payOut(PayoutRequest request, StreamObserver<PayoutResponse> responseObserver) {
        log.info("Received payOut request: payment_id={}, currency={}, amount={}",
                request.getPaymentId(),
                request.getCurrency(),
                request.getAmount());

        // TODO: Implement your payout logic here
        // 1. Validate the payout request
        // 2. Initiate the payout in your system
        // 3. Return the appropriate response

        // Example: Accept the payout
        responseObserver.onNext(PayoutResponse.newBuilder()
                .setAccepted(PayoutResponse.Accepted.newBuilder().build())
                .build());
        responseObserver.onCompleted();

        // TODO: finalizePayout should be called when your system completes (or fails) the payout
        new Thread(() -> {
            try {
                Thread.sleep(2000);
                networkClient.finalizePayout(FinalizePayoutRequest.newBuilder()
                        .setPaymentId(request.getPaymentId())
                        .setSuccess(FinalizePayoutRequest.Success.newBuilder()
                                .setReceipt(PaymentReceipt.newBuilder()
                                        .setSepa(PaymentReceipt.Sepa.newBuilder()
                                                .setBankingTransactionReferenceId("1234567890")
                                                .build())
                                        .build())
                                .build())
                        .build());
            } catch (Exception e) {
                log.error("Failed to finalize payout for payment {}", request.getPaymentId(), e);
            }
        }).start();
    }

    @Override
    public void updateLimit(UpdateLimitRequest request, StreamObserver<UpdateLimitResponse> responseObserver) {
        // TODO: optionally implement handling of the notifications about updates on your limits and limits usage
        log.info("Received updateLimit with {} limits", request.getLimitsCount());

        for (UpdateLimitRequest.Limit limit : request.getLimitsList()) {
            log.debug("Limit update: counterpart_id={}, payout_limit={}, credit_limit={}, credit_usage={}",
                    limit.getCounterpartId(),
                    limit.getPayoutLimit(),
                    limit.getCreditLimit(),
                    limit.getCreditUsage());
        }

        responseObserver.onNext(UpdateLimitResponse.newBuilder().build());
        responseObserver.onCompleted();
    }

    @Override
    public void appendLedgerEntries(AppendLedgerEntriesRequest request, StreamObserver<AppendLedgerEntriesResponse> responseObserver) {
        // TODO: optionally implement handling of the notifications about new ledger transactions and new ledger entries
        log.info("Received appendLedgerEntries with {} transactions", request.getTransactionsCount());

        for (AppendLedgerEntriesRequest.Transaction transaction : request.getTransactionsList()) {
            log.debug("Transaction: id={}, entries_count={}",
                    transaction.getTransactionId(),
                    transaction.getEntriesCount());
        }

        responseObserver.onNext(AppendLedgerEntriesResponse.newBuilder().build());
        responseObserver.onCompleted();
    }

    @Override
    public void approvePaymentQuotes(ApprovePaymentQuoteRequest request, StreamObserver<ApprovePaymentQuoteResponse> responseObserver) {
        log.info("Received approvePaymentQuotes for payment_id={}", request.getPaymentId());

        // TODO: Implement "Last Look" logic
        // Verify that the final rates and amounts are acceptable

        // Example: Accept the quotes
        responseObserver.onNext(ApprovePaymentQuoteResponse.newBuilder()
                .setAccepted(ApprovePaymentQuoteResponse.Accepted.newBuilder().build())
                .build());
        responseObserver.onCompleted();
    }
}

File: csharp/starter/template/Services/PaymentHandler.cs
using Grpc.Core;
using T0.ProviderSdk.Api.Tzero.V1.Common;
using T0.ProviderSdk.Api.Tzero.V1.Payment;

namespace {{PROJECT_NAME_PASCAL}}.Services;

// Please refer to docs and proto definitions to understand the purpose of each method.
public class PaymentHandler(
    NetworkService.NetworkServiceClient networkClient,
    ILogger<PaymentHandler> logger) : ProviderService.ProviderServiceBase
{
    // TODO: Step 2.1 Implement how you handle updates of payments initiated by you
    public override Task<UpdatePaymentResponse> UpdatePayment(
        UpdatePaymentRequest request, ServerCallContext context)
    {
        logger.LogInformation("UpdatePayment: payment_id={PaymentId}", request.PaymentId);
        return Task.FromResult(new UpdatePaymentResponse());
    }

    // TODO: Step 2.4 Implement how you do payouts (payments initiated by your counterparts)
    public override async Task<PayoutResponse> PayOut(
        PayoutRequest request, ServerCallContext context)
    {
        logger.LogInformation("PayOut: payment_id={PaymentId}, currency={Currency}",
            request.PaymentId, request.Currency);

        // optional: if this payment needs a manual AML check on your side, respond with
        // ManualAmlCheck here, before making the payout, and report the outcome later via
        // CompleteManualAmlCheck (see Services/CompleteManualAmlCheck.cs). Do not finalize on
        // this path — the payout only proceeds once the check is approved.
        // return new PayoutResponse { ManualAmlCheck = new PayoutResponse.Types.ManualAmlCheck() };

        // TODO: FinalizePayout should be called when your system completes the payout
        await networkClient.FinalizePayoutAsync(new FinalizePayoutRequest
        {
            PaymentId = request.PaymentId,
            Success = new FinalizePayoutRequest.Types.Success
            {
                Receipt = new PaymentReceipt
                {
                    Sepa = new PaymentReceipt.Types.Sepa
                    {
                        BankingTransactionReferenceId = "123456"
                    }
                }
            }
        });

        // optional: if your provider has multiple legal entities, set BeneficiaryProviderLegalEntityId
        return new PayoutResponse { Accepted = new PayoutResponse.Types.Accepted() };
    }

    // TODO: Optionally implement handling of limit update notifications
    public override Task<UpdateLimitResponse> UpdateLimit(
        UpdateLimitRequest request, ServerCallContext context)
    {
        return Task.FromResult(new UpdateLimitResponse());
    }

    // TODO: Optionally implement handling of new ledger transactions
    public override Task<AppendLedgerEntriesResponse> AppendLedgerEntries(
        AppendLedgerEntriesRequest request, ServerCallContext context)
    {
        return Task.FromResult(new AppendLedgerEntriesResponse());
    }

    // TODO: Implement "Last Look" — verify final rates and approve after AML check.
    // The request includes PayOutFix — the fixed charge in USD for this payout.
    // Consider it alongside PayOutRate and PayOutAmount when deciding to accept.
    public override Task<ApprovePaymentQuoteResponse> ApprovePaymentQuotes(
        ApprovePaymentQuoteRequest request, ServerCallContext context)
    {
        return Task.FromResult(new ApprovePaymentQuoteResponse
        {
            Accepted = new ApprovePaymentQuoteResponse.Types.Accepted()
        });
    }
}
